lapce-data/src/hover.rs

```diff
@@ -152,41 +170,317 @@ impl Default for HoverData {
+    }
+
+    fn links(&self) -> &[druid::text::Link] {
+        self.text.links()
+    }
+}
+
+// TODO: This could moved into a more general utility file, for use in things other than hover
+/// Parse the given markdown into renderable druid rich text
+/// `base_font_size` is the font size of normal text, and the size of headers is computed from it
+fn parse_markdown(text: &str, style: HoverTextStyle) -> MarkdownText {
+    use pulldown_cmark::{Event, Options, Parser};
+
+    let mut builder = RichTextBuilder::new();
+
+    // Our position within the text
+    let mut pos = 0;
+
+    // TODO: (minor): This could use a smallvec since most tags are probably not that nested
+    // Stores the current tags (like italics/bold/strikethrough) so that they can be nested
+    let mut tag_stack = Vec::new();
+
+    let mut code_block_indices = Vec::new();
+
+    // Construct the markdown parser. We enable most of the options in order to provide the most
+    // compatibility that pulldown_cmark allows.
+    let parser = Parser::new_ext(
+        text,
+        Options::ENABLE_TABLES
+            | Options::ENABLE_FOOTNOTES
+            | Options::ENABLE_STRIKETHROUGH
+            | Options::ENABLE_TASKLISTS
+            | Options::ENABLE_HEADING_ATTRIBUTES,
+    );
+    for event in parser {
+        match event {
+            Event::Start(tag) => {
+                tag_stack.push((pos, tag));
+            }
+            Event::End(end_tag) => {
+                if let Some((start_offset, tag)) = tag_stack.pop() {
+                    if end_tag != tag {
+                        log::warn!("Mismatched markdown tag");
+                        continue;
+                    }
+
+                    if let Tag::CodeBlock(_kind) = &tag {
+                        code_block_indices.push(start_offset..pos);
+                    }
+
+                    add_attribute_for_tag(
+                        &tag,
+                        builder.add_attributes_for_range(start_offset..pos),
+                        style.clone(),
+                    );
+
+                    if should_add_newline_after_tag(&tag) {
+                        builder.push("\n");
+                        pos += 1;
+                    }
+                } else {
+                    // Very minor warning. We got a bit of malformed markdown (according to
+                    // pulldown_cmark, at least)
+                    log::warn!("Unbalanced markdown tag")
+                }
+            }
+            Event::Text(text) => {
+                builder.push(&text);
+                pos += text.len();
+            }
+            Event::Code(text) => {
+                builder.push(&text).font_family(FontFamily::MONOSPACE);
+                code_block_indices.push(pos..(pos + text.len()));
+                pos += text.len();
+            }
+            // TODO: Some minimal 'parsing' of html could be useful here, since some things use
+            // basic html like `<code>text</code>`.
```

Comment:
I think we should be consistent with what rustdoc does here. But I understand this is the most significant simplification made in this PR, and for good reasons.

Reply:
I included that because some of the rust documentation actually uses `<code>stuff</code>`, for some reason.